internal/downloader/s3.go

```diff
@@ -0,0 +1,49 @@
+package downloader
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/pkg/errors"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+)
+
+type downloaderAPI interface {
+	Download(w io.WriterAt, input *s3.GetObjectInput, options ...func(*s3manager.Downloader)) (n int64, err error)
+	DownloadWithContext(ctx aws.Context, w io.WriterAt, input *s3.GetObjectInput, options ...func(*s3manager.Downloader)) (n int64, err error)
+	DownloadWithIterator(ctx aws.Context, iter s3manager.BatchDownloadIterator, opts ...func(*s3manager.Downloader)) error
+}
+
+type Downloader struct {
+	Bucket string
+	d      downloaderAPI
+}
+
+func New(bucketName string, dl downloaderAPI) (*Downloader, error) {
+	if dl == nil {
+		return nil, fmt.Errorf("received nil downloader")
+	}
+
+	return &Downloader{
+		Bucket: bucketName,
+		d:      dl,
+	}, nil
+}
+
+func (dl *Downloader) Download(key string) ([]byte, error) {
```

Comment:
We could consider extending this to provide a more complete abstraction. Callers will always want to marshal these bytes into a `*chart.Chart` so they can actually do something with it. We could encapsulate this behaviour in a `S3ChartDownloader` (as an instance of a hypothetical `ChartDownloader` interface defined elsewhere) and provide a more readily usable abstraction.

Reply:
I have added and tested a function which returns the `*chart.Chart` type. As you stated, a general `ChartDownloader` interface can be defined elsewhere if someone would like to use a different chart repository from S3.